lightning-invoice/src/payment.rs

```diff
@@ -462,26 +466,30 @@ where
 	fn handle_event(&self, event: &Event) {
 		match event {
 			Event::PaymentPathFailed {
-				all_paths_failed, payment_id, payment_hash, rejected_by_dest, path,
-				short_channel_id, retry, ..
+				payment_id, payment_hash, rejected_by_dest, path, short_channel_id, retry, ..
 			} => {
 				if let Some(short_channel_id) = short_channel_id {
 					let path = path.iter().collect::<Vec<_>>();
 					self.scorer.lock().payment_path_failed(&path, *short_channel_id);
 				}
 
-				if *rejected_by_dest {
-					log_trace!(self.logger, "Payment {} rejected by destination; not retrying", log_bytes!(payment_hash.0));
-				} else if payment_id.is_none() {
+				if payment_id.is_none() {
 					log_trace!(self.logger, "Payment {} has no id; not retrying", log_bytes!(payment_hash.0));
+				} else if *rejected_by_dest {
+					log_trace!(self.logger, "Payment {} rejected by destination; not retrying", log_bytes!(payment_hash.0));
+					self.payer.no_further_payment_retries(payment_id.unwrap());
 				} else if retry.is_none() {
 					log_trace!(self.logger, "Payment {} missing retry params; not retrying", log_bytes!(payment_hash.0));
+					self.payer.no_further_payment_retries(payment_id.unwrap());
```

Comment:
Could these two cases be handled directly inside `ChannelManager` such that `no_further_payment_retries` only needs to be called when `retry_payment` returns an error?

Reply:
Hmm, `retry.is_none` is just a backwards compatibility thing - we'll remove it sooner or later - but I guess in this context `no_further_payment_retries` is actually a noop, just feels nice to have it for consistency? As for `rejected_by_dest`, I don't think we want to do that - in the case of someone implementing payment probes users may want to retry other failing paths even if some paths fail with `rejected_by_dest`. We don't really support that here, I guess, but its totally reasonable for a user to want to have their own `InvoicePayer` that uses it.

Reply:
You can't use `InvoicePayer` with per-path retries and probing anyway (well, mostly cause `InvoicePayer`'s per-path retries are broken entirely), so I don't think anything new broke here?

Reply:
Right, my point was more that, prior to this PR, retries for individual paths were broken in general (ie cause they get retried more after retries are up). For a non-probe payment, it feels "obvious" that a `rejected_by_dest` failure should cause us to stop the payment - we won't be able to complete the payment anyway, so we should stop. For a probe, indeed, indeed, we want different behavior, but I'm not sure how to expose that here...the new ChannelManager API suggested is focused on informing us either when we get a single success or when all payments have failed after we tell it we're done. What we need for probes the ability to learn when all paths have failed even though we don't want to mark the payment as done until all paths have failed with some condition.

Trying to accomplish this without explicit ChannelManager support for it seems to keep bumping up against #1164 again - we can't know just by looking at the event in front of us if its the last one for a given payment or not. If its tagged `rejected_by_dest` we can skip retrying and expose it to users, but then we've just pushed the "is the payment complete" detection to our users, where there isn't a clear way to do it. We could try some fuzzy heuristic on the amount of funds which have "successfully failed", but that fails to account for fees in some cases (plus doesn't work across restarts), so I'm kinda hesitant to do it. Maybe if we don't care about the fees issue (since we're just collecting paths) and consider it acceptable that it fails across restarts (cause its just a probe), then we can add some new API in the `InvoicePayer` to explicitly probe with some path-collection and then generate a new event there that says "collected enough paths, here is a route"? We'll need to be really careful about re-using liquidity twice in that case, though, which may make things even more complicated.